File: app/src/main/java/com/kaoyan/wordhelper/util/Sm2Scheduler.kt
package com.kaoyan.wordhelper.util

import com.kaoyan.wordhelper.data.entity.Progress
import com.kaoyan.wordhelper.data.model.SpellingOutcome
import com.kaoyan.wordhelper.data.model.StudyRating
import java.time.Instant
import java.time.ZoneId
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.roundToInt

object Sm2Scheduler {

    private const val DEFAULT_EASE = 2.5f
    private const val MIN_EASE = 1.3f
    private const val MIN_EASE_SPELLING = 1.1f
    private const val MAX_EASE = 3.0f
    private const val ONE_MINUTE = 60 * 1000L
    private const val TEN_MINUTES = 10 * 60 * 1000L
    private const val DAY_REFRESH_HOUR = 4
    private const val MASTERED_INTERVAL_DAYS_V3 = 21
    private const val MASTERED_MIN_REVIEW_COUNT_V3 = 2
    private const val MASTERED_INTERVAL_DAYS_V4 = 30
    private const val MASTERED_MIN_REVIEW_COUNT_V4 = 4
    private const val MASTERED_MIN_EASE = 2.3f
    private const val MAX_GROWTH_FACTOR = 2.5f
    private const val HARD_GROWTH_FACTOR = 1.2f
    private const val SHORT_TERM_REVIEW_DECAY_HOURS = 12L
    private const val V3_MIN_INTERVAL_DAYS = 1
    private const val V3_GOOD_INTERVAL_DAYS = 2
    private val EBBINGHAUS_LADDER = intArrayOf(1, 2, 6, 14, 30)

    data class ScheduleResult(
        val repetitions: Int,
        val intervalDays: Int,
        val easeFactor: Float,
        val nextReviewTime: Long,
        val status: Int
    )

    fun schedule(
        current: Progress?,
        rating: StudyRating,
        now: Long = System.currentTimeMillis(),
        algorithmV4Enabled: Boolean = false
    ): ScheduleResult {
        if (algorithmV4Enabled) {
            return scheduleRecognitionV4(current = current, rating = rating, now = now)
        }

        var ease = updateEase(current?.easeFactor ?: DEFAULT_EASE, rating.quality, MIN_EASE)
        val previousRepetitions = current?.repetitions ?: 0
        val previousIntervalDays = current?.intervalDays ?: 0
        val previousReviewCount = current?.reviewCount ?: 0
        val isLearningPhase = isLearningPhase(current)

        if (rating == StudyRating.AGAIN) {
            ease = (ease - 0.2f).coerceAtLeast(MIN_EASE)
            return ScheduleResult(
                repetitions = 0,
                intervalDays = V3_MIN_INTERVAL_DAYS,
                easeFactor = ease,
                nextReviewTime = computeNextReviewTimeByDays(now, V3_MIN_INTERVAL_DAYS),
                status = Progress.STATUS_LEARNING
            )
        }

        if (isLearningPhase) {
            val intervalDays = when (rating) {
                StudyRating.HARD -> V3_MIN_INTERVAL_DAYS
                StudyRating.GOOD -> V3_GOOD_INTERVAL_DAYS
                StudyRating.AGAIN -> V3_MIN_INTERVAL_DAYS
            }
            val effectiveReviewCount = previousReviewCount + 1
            return ScheduleResult(
                repetitions = previousRepetitions + 1,
                intervalDays = intervalDays,
                easeFactor = ease,
                nextReviewTime = nextReviewTimeByInterval(now, intervalDays),
                status = resolveStatus(intervalDays, effectiveReviewCount, ease, algorithmV4Enabled = false)
            )
        }

        val oldInterval = previousIntervalDays.coerceAtLeast(1)
        val intervalDays = if (rating == StudyRating.HARD) {
            conservativeInterval(oldInterval, ease)
        } else {
            standardGoodInterval(oldInterval, ease)
        }

        val effectiveReviewCount = previousReviewCount + 1
        val repetitions = previousRepetitions + 1
        val status = resolveStatus(intervalDays, effectiveReviewCount, ease, algorithmV4Enabled = false)
        val nextReviewTime = computeNextReviewTimeByDays(now, intervalDays)

        return ScheduleResult(
            repetitions = repetitions,
            intervalDays = intervalDays,
            easeFactor = ease,
            nextReviewTime = nextReviewTime,
            status = status
        )
    }

    private fun scheduleRecognitionV4(
        current: Progress?,
        rating: StudyRating,
        now: Long
    ): ScheduleResult {
        var ease = updateEase(current?.easeFactor ?: DEFAULT_EASE, rating.quality, MIN_EASE)
        val previousRepetitions = current?.repetitions ?: 0
        val previousIntervalDays = current?.intervalDays ?: 0
        val previousReviewCount = current?.reviewCount ?: 0
        val isLearningPhase = isLearningPhase(current)

        if (rating == StudyRating.AGAIN) {
            ease = (ease - 0.2f).coerceAtLeast(MIN_EASE)
            return ScheduleResult(
                repetitions = 0,
                intervalDays = 0,
                easeFactor = ease,
                nextReviewTime = now + ONE_MINUTE,
                status = Progress.STATUS_LEARNING
            )
        }

        var intervalDays = if (isLearningPhase) {
            when (rating) {
                StudyRating.HARD -> 0
                StudyRating.GOOD -> nextEbbinghausStep(previousIntervalDays)
                StudyRating.AGAIN -> 0
            }
        } else {
            val oldInterval = previousIntervalDays.coerceAtLeast(1)
            val rawInterval = if (rating == StudyRating.HARD) {
                conservativeInterval(oldInterval, ease)
            } else {
                standardGoodInterval(oldInterval, ease)
            }
            closestEbbinghausStep(rawInterval)
        }

        if (intervalDays > 1 && reviewedWithinHours(current, now, SHORT_TERM_REVIEW_DECAY_HOURS)) {
            intervalDays = 1
        }

        val effectiveReviewCount = previousReviewCount + 1
        val repetitions = previousRepetitions + 1
        val status = resolveStatus(intervalDays, effectiveReviewCount, ease, algorithmV4Enabled = true)

        return ScheduleResult(
            repetitions = repetitions,
            intervalDays = intervalDays,
            easeFactor = ease,
            nextReviewTime = nextReviewTimeByInterval(now, intervalDays),
            status = status
        )
    }

    fun scheduleSpelling(
        current: Progress?,
        outcome: SpellingOutcome,
        now: Long = System.currentTimeMillis(),
        algorithmV4Enabled: Boolean = false
    ): ScheduleResult {
        if (algorithmV4Enabled) {
            return scheduleSpellingV4(current = current, outcome = outcome, now = now)
        }

        val previousReviewCount = current?.reviewCount ?: 0
        val previousIntervalDays = current?.intervalDays ?: 0
        val previousRepetitions = current?.repetitions ?: 0
        val isLearningPhase = isLearningPhase(current)
        var ease = updateEase(current?.easeFactor ?: DEFAULT_EASE, outcome.quality, MIN_EASE_SPELLING)

        if (outcome == SpellingOutcome.FAILED) {
            ease = (ease - 0.2f).coerceAtLeast(MIN_EASE_SPELLING)
            return ScheduleResult(
                repetitions = 0,
                intervalDays = V3_MIN_INTERVAL_DAYS,
                easeFactor = ease,
                nextReviewTime = computeNextReviewTimeByDays(now, V3_MIN_INTERVAL_DAYS),
                status = Progress.STATUS_LEARNING
            )
        }

        val intervalDays = when {
            isLearningPhase -> {
                when (outcome) {
                    SpellingOutcome.RETRY_SUCCESS -> V3_MIN_INTERVAL_DAYS
                    SpellingOutcome.HINTED,
                    SpellingOutcome.PERFECT -> V3_GOOD_INTERVAL_DAYS
                    SpellingOutcome.FAILED -> V3_MIN_INTERVAL_DAYS
                }
            }

            outcome == SpellingOutcome.RETRY_SUCCESS -> {
                conservativeInterval(previousIntervalDays.coerceAtLeast(1), ease)
            }

            outcome == SpellingOutcome.HINTED -> {
                standardGoodInterval(previousIntervalDays.coerceAtLeast(1), ease)
            }

            else -> {
                val base = standardGoodInterval(previousIntervalDays.coerceAtLeast(1), ease)
                (base * 1.1f).roundToInt().coerceAtLeast(1)
            }
        }
        val effectiveReviewCount = previousReviewCount + 1
        val repetitions = previousRepetitions + 1
        val status = resolveStatus(intervalDays, effectiveReviewCount, ease, algorithmV4Enabled = false)

        return ScheduleResult(
            repetitions = repetitions,
            intervalDays = intervalDays,
            easeFactor = ease,
            nextReviewTime = nextReviewTimeByInterval(now, intervalDays),
            status = status
        )
    }

    private fun scheduleSpellingV4(
        current: Progress?,
        outcome: SpellingOutcome,
        now: Long
    ): ScheduleResult {
        val previousReviewCount = current?.reviewCount ?: 0
        val previousIntervalDays = current?.intervalDays ?: 0
        val previousRepetitions = current?.repetitions ?: 0
        val isLearningPhase = isLearningPhase(current)
        var ease = updateEase(current?.easeFactor ?: DEFAULT_EASE, outcome.quality, MIN_EASE_SPELLING)

        if (outcome == SpellingOutcome.FAILED) {
            ease = (ease - 0.2f).coerceAtLeast(MIN_EASE_SPELLING)
            return ScheduleResult(
                repetitions = 0,
                intervalDays = 0,
                easeFactor = ease,
                nextReviewTime = now + ONE_MINUTE,
                status = Progress.STATUS_LEARNING
            )
        }

        var intervalDays = if (isLearningPhase) {
            when (outcome) {
                SpellingOutcome.RETRY_SUCCESS -> 0
                SpellingOutcome.HINTED,
                SpellingOutcome.PERFECT -> nextEbbinghausStep(previousIntervalDays)
                SpellingOutcome.FAILED -> 0
            }
        } else {
            val oldInterval = previousIntervalDays.coerceAtLeast(1)
            val rawInterval = when (outcome) {
                SpellingOutcome.RETRY_SUCCESS -> {
                    conservativeInterval(oldInterval, ease)
                }

                SpellingOutcome.HINTED -> {
                    standardGoodInterval(oldInterval, ease)
                }

                SpellingOutcome.PERFECT -> {
                    val base = standardGoodInterval(oldInterval, ease)
                    (base * 1.1f).roundToInt().coerceAtLeast(1)
                }

                SpellingOutcome.FAILED -> 1
            }
            closestEbbinghausStep(rawInterval)
        }

        if (intervalDays > 1 && reviewedWithinHours(current, now, SHORT_TERM_REVIEW_DECAY_HOURS)) {
            intervalDays = 1
        }

        val effectiveReviewCount = previousReviewCount + 1
        val repetitions = previousRepetitions + 1
        val status = resolveStatus(intervalDays, effectiveReviewCount, ease, algorithmV4Enabled = true)

        return ScheduleResult(
            repetitions = repetitions,
            intervalDays = intervalDays,
            easeFactor = ease,
            nextReviewTime = nextReviewTimeByInterval(now, intervalDays),
            status = status
        )
    }

    private fun isLearningPhase(current: Progress?): Boolean {
        if (current == null) return true
        if (current.status == Progress.STATUS_NEW) return true
        return current.intervalDays <= 0 || current.repetitions <= 0
    }

    private fun updateEase(currentEase: Float, quality: Int, minEase: Float): Float {
        val updated = currentEase - 0.8f + 0.28f * quality - 0.02f * quality * quality
        return updated.coerceIn(minEase, MAX_EASE)
    }

    private fun conservativeInterval(oldIntervalDays: Int, ease: Float): Int {
        val safeOldInterval = oldIntervalDays.coerceAtLeast(1)
        val byHard = (safeOldInterval * HARD_GROWTH_FACTOR).roundToInt()
        val byEase = (safeOldInterval * ease).roundToInt()
        return min(byHard, byEase).coerceAtLeast(1)
    }

    private fun standardGoodInterval(oldIntervalDays: Int, ease: Float): Int {
        val safeOldInterval = oldIntervalDays.coerceAtLeast(1)
        val grown = (safeOldInterval * ease).roundToInt().coerceAtLeast(1)
        val maxAllowed = (safeOldInterval * MAX_GROWTH_FACTOR).roundToInt().coerceAtLeast(1)
        return min(grown, maxAllowed)
    }

    private fun reviewedWithinHours(current: Progress?, now: Long, hours: Long): Boolean {
        val estimatedLastReview = estimateLastReviewTime(current) ?: return false
        val elapsed = now - estimatedLastReview
        if (elapsed < 0L) return false
        val threshold = hours * 60 * 60 * 1000
        return elapsed <= threshold
    }

    private fun nextEbbinghausStep(currentIntervalDays: Int): Int {
        if (currentIntervalDays <= 0) return EBBINGHAUS_LADDER.first()
        return EBBINGHAUS_LADDER.firstOrNull { it > currentIntervalDays } ?: EBBINGHAUS_LADDER.last()
    }

    private fun closestEbbinghausStep(intervalDays: Int): Int {
        val safeInterval = intervalDays.coerceAtLeast(1)
        var nearest = EBBINGHAUS_LADDER.first()
        EBBINGHAUS_LADDER.forEach { step ->
            val stepDiff = abs(step - safeInterval)
            val nearestDiff = abs(nearest - safeInterval)
            if (stepDiff < nearestDiff || (stepDiff == nearestDiff && step > nearest)) {
                nearest = step
            }
        }
        return nearest
    }

    private fun resolveStatus(
        intervalDays: Int,
        reviewCount: Int,
        easeFactor: Float,
        algorithmV4Enabled: Boolean
    ): Int {
        val masteredIntervalDays = if (algorithmV4Enabled) {
            MASTERED_INTERVAL_DAYS_V4
        } else {
            MASTERED_INTERVAL_DAYS_V3
        }
        val masteredMinReviewCount = if (algorithmV4Enabled) {
            MASTERED_MIN_REVIEW_COUNT_V4
        } else {
            MASTERED_MIN_REVIEW_COUNT_V3
        }
        return if (
            intervalDays >= masteredIntervalDays &&
            reviewCount >= masteredMinReviewCount &&
            easeFactor >= MASTERED_MIN_EASE
        ) {
            Progress.STATUS_MASTERED
        } else {
            Progress.STATUS_LEARNING
        }
    }

    private fun nextReviewTimeByInterval(now: Long, intervalDays: Int): Long {
        return if (intervalDays > 0) {
            computeNextReviewTimeByDays(now, intervalDays)
        } else {
            now + TEN_MINUTES
        }
    }

    fun estimateLastReviewTime(progress: Progress?): Long? {
        if (progress == null || progress.reviewCount <= 0 || progress.nextReviewTime <= 0L) {
            return null
        }
        return if (progress.intervalDays > 0) {
            Instant.ofEpochMilli(progress.nextReviewTime)
                .atZone(ZoneId.systemDefault())
                .minusDays(progress.intervalDays.toLong())
                .toInstant()
                .toEpochMilli()
                .coerceAtLeast(0L)
        } else {
            (progress.nextReviewTime - TEN_MINUTES).coerceAtLeast(0L)
        }
    }

    fun nextReviewTimeByDays(intervalDays: Int, now: Long = System.currentTimeMillis()): Long {
        return computeNextReviewTimeByDays(now, intervalDays)
    }

    private fun computeNextReviewTimeByDays(now: Long, intervalDays: Int): Long {
        val safeDays = intervalDays.coerceAtLeast(1).toLong()
        val zoneId = ZoneId.systemDefault()
        val zonedNow = Instant.ofEpochMilli(now).atZone(zoneId)
        val learningDate = if (zonedNow.hour < DAY_REFRESH_HOUR) {
            zonedNow.toLocalDate().minusDays(1)
        } else {
            zonedNow.toLocalDate()
        }
        return learningDate
            .plusDays(safeDays)
            .atTime(DAY_REFRESH_HOUR, 0)
            .atZone(zoneId)
            .toInstant()
            .toEpochMilli()
    }
}

File: web/src/utils/sm2Scheduler.ts
import {
  PROGRESS_STATUS_LEARNING,
  PROGRESS_STATUS_MASTERED,
  SPELLING_OUTCOME_FAILED,
  SPELLING_OUTCOME_HINTED,
  SPELLING_OUTCOME_PERFECT,
  SPELLING_OUTCOME_RETRY_SUCCESS,
  STUDY_RATING_AGAIN,
  STUDY_RATING_GOOD,
  STUDY_RATING_HARD,
} from "../types";
import type { Progress, SpellingOutcome, StudyRating } from "../types";
import { nextReviewTimeByDays } from "./dateUtils";

const DEFAULT_EASE = 2.5;
const MIN_EASE = 1.3;
const MIN_EASE_SPELLING = 1.1;
const MAX_EASE = 3.0;
const ONE_MINUTE = 60 * 1000;
const TEN_MINUTES = 10 * 60 * 1000;
const MASTERED_INTERVAL_DAYS_V3 = 21;
const MASTERED_MIN_REVIEW_COUNT_V3 = 2;
const MASTERED_INTERVAL_DAYS_V4 = 30;
const MASTERED_MIN_REVIEW_COUNT_V4 = 4;
const MASTERED_MIN_EASE = 2.3;
const MAX_GROWTH_FACTOR = 2.5;
const HARD_GROWTH_FACTOR = 1.2;
const SHORT_TERM_REVIEW_DECAY_HOURS = 12;
const V3_MIN_INTERVAL_DAYS = 1;
const V3_GOOD_INTERVAL_DAYS = 2;
const EBBINGHAUS_LADDER = [1, 2, 6, 14, 30];

export interface ScheduleResult {
  repetitions: number;
  intervalDays: number;
  easeFactor: number;
  nextReviewTime: number;
  status: 1 | 2;
}

const updateEase = (currentEase: number, quality: number, minEase: number): number => {
  const updated = currentEase - 0.8 + 0.28 * quality - 0.02 * quality * quality;
  return Math.max(minEase, Math.min(MAX_EASE, updated));
};

const conservativeInterval = (oldInterval: number, ease: number): number => {
  const safeOld = Math.max(1, oldInterval);
  const byHard = Math.round(safeOld * HARD_GROWTH_FACTOR);
  const byEase = Math.round(safeOld * ease);
  return Math.max(1, Math.min(byHard, byEase));
};

const standardGoodInterval = (oldInterval: number, ease: number): number => {
  const safeOld = Math.max(1, oldInterval);
  const grown = Math.max(1, Math.round(safeOld * ease));
  const maxAllowed = Math.max(1, Math.round(safeOld * MAX_GROWTH_FACTOR));
  return Math.min(grown, maxAllowed);
};

const isLearningPhase = (current: Progress | null): boolean => {
  if (!current) return true;
  return current.intervalDays <= 0 || current.repetitions <= 0;
};

const resolveStatus = (
  intervalDays: number,
  reviewCount: number,
  easeFactor: number,
  algorithmV4Enabled: boolean,
): 1 | 2 => {
  const masteredIntervalDays = algorithmV4Enabled ? MASTERED_INTERVAL_DAYS_V4 : MASTERED_INTERVAL_DAYS_V3;
  const masteredReviewCount = algorithmV4Enabled ? MASTERED_MIN_REVIEW_COUNT_V4 : MASTERED_MIN_REVIEW_COUNT_V3;

  if (
    intervalDays >= masteredIntervalDays &&
    reviewCount >= masteredReviewCount &&
    easeFactor >= MASTERED_MIN_EASE
  ) {
    return PROGRESS_STATUS_MASTERED;
  }
  return PROGRESS_STATUS_LEARNING;
};

const estimateLastReview = (current: Progress | null, now: number): number | null => {
  if (!current || current.reviewCount <= 0 || current.nextReviewTime <= 0) return null;
  if (current.lastReviewTime > 0) return current.lastReviewTime;
  if (current.intervalDays > 0) {
    return current.nextReviewTime - current.intervalDays * 24 * 60 * 60 * 1000;
  }
  return now - TEN_MINUTES;
};

const reviewedWithinHours = (current: Progress | null, now: number, hours: number): boolean => {
  const estimated = estimateLastReview(current, now);
  if (!estimated) return false;
  const elapsed = now - estimated;
  if (elapsed < 0) return false;
  return elapsed <= hours * 60 * 60 * 1000;
};

const nextEbbinghausStep = (currentIntervalDays: number): number => {
  if (currentIntervalDays <= 0) return EBBINGHAUS_LADDER[0];
  return EBBINGHAUS_LADDER.find((step) => step > currentIntervalDays) ?? EBBINGHAUS_LADDER[EBBINGHAUS_LADDER.length - 1];
};

const closestEbbinghausStep = (intervalDays: number): number => {
  const safe = Math.max(1, intervalDays);
  let nearest = EBBINGHAUS_LADDER[0];
  for (const step of EBBINGHAUS_LADDER) {
    const diff = Math.abs(step - safe);
    const nearestDiff = Math.abs(nearest - safe);
    if (diff < nearestDiff || (diff === nearestDiff && step > nearest)) {
      nearest = step;
    }
  }
  return nearest;
};

const nextReviewTimeByInterval = (now: number, intervalDays: number): number => {
  if (intervalDays > 0) {
    return nextReviewTimeByDays(intervalDays, now);
  }
  return now + TEN_MINUTES;
};

export const scheduleRecognition = (
  current: Progress | null,
  rating: StudyRating,
  now: number = Date.now(),
  algorithmV4Enabled: boolean = false,
): ScheduleResult => {
  if (algorithmV4Enabled) {
    return scheduleRecognitionV4(current, rating, now);
  }

  let ease = updateEase(current?.easeFactor ?? DEFAULT_EASE, rating, MIN_EASE);
  const prevRepetitions = current?.repetitions ?? 0;
  const prevIntervalDays = current?.intervalDays ?? 0;
  const prevReviewCount = current?.reviewCount ?? 0;
  const learningPhase = isLearningPhase(current);

  if (rating === STUDY_RATING_AGAIN) {
    ease = Math.max(MIN_EASE, ease - 0.2);
    return {
      repetitions: 0,
      intervalDays: V3_MIN_INTERVAL_DAYS,
      easeFactor: ease,
      nextReviewTime: nextReviewTimeByDays(V3_MIN_INTERVAL_DAYS, now),
      status: PROGRESS_STATUS_LEARNING,
    };
  }

  if (learningPhase) {
    const intervalDays = rating === STUDY_RATING_HARD ? V3_MIN_INTERVAL_DAYS : V3_GOOD_INTERVAL_DAYS;
    const effectiveReviewCount = prevReviewCount + 1;
    return {
      repetitions: prevRepetitions + 1,
      intervalDays,
      easeFactor: ease,
      nextReviewTime: nextReviewTimeByInterval(now, intervalDays),
      status: resolveStatus(intervalDays, effectiveReviewCount, ease, false),
    };
  }

  const oldInterval = Math.max(1, prevIntervalDays);
  const intervalDays =
    rating === STUDY_RATING_HARD ? conservativeInterval(oldInterval, ease) : standardGoodInterval(oldInterval, ease);

  const effectiveReviewCount = prevReviewCount + 1;
  return {
    repetitions: prevRepetitions + 1,
    intervalDays,
    easeFactor: ease,
    nextReviewTime: nextReviewTimeByDays(intervalDays, now),
    status: resolveStatus(intervalDays, effectiveReviewCount, ease, false),
  };
};

const scheduleRecognitionV4 = (current: Progress | null, rating: StudyRating, now: number): ScheduleResult => {
  let ease = updateEase(current?.easeFactor ?? DEFAULT_EASE, rating, MIN_EASE);
  const prevRepetitions = current?.repetitions ?? 0;
  const prevIntervalDays = current?.intervalDays ?? 0;
  const prevReviewCount = current?.reviewCount ?? 0;
  const learningPhase = isLearningPhase(current);

  if (rating === STUDY_RATING_AGAIN) {
    ease = Math.max(MIN_EASE, ease - 0.2);
    return {
      repetitions: 0,
      intervalDays: 0,
      easeFactor: ease,
      nextReviewTime: now + ONE_MINUTE,
      status: PROGRESS_STATUS_LEARNING,
    };
  }

  let intervalDays = learningPhase
    ? rating === STUDY_RATING_HARD
      ? 0
      : nextEbbinghausStep(prevIntervalDays)
    : closestEbbinghausStep(
        rating === STUDY_RATING_HARD
          ? conservativeInterval(Math.max(1, prevIntervalDays), ease)
          : standardGoodInterval(Math.max(1, prevIntervalDays), ease),
      );

  if (intervalDays > 1 && reviewedWithinHours(current, now, SHORT_TERM_REVIEW_DECAY_HOURS)) {
    intervalDays = 1;
  }

  const effectiveReviewCount = prevReviewCount + 1;
  return {
    repetitions: prevRepetitions + 1,
    intervalDays,
    easeFactor: ease,
    nextReviewTime: nextReviewTimeByInterval(now, intervalDays),
    status: resolveStatus(intervalDays, effectiveReviewCount, ease, true),
  };
};

export const scheduleSpelling = (
  current: Progress | null,
  outcome: SpellingOutcome,
  now: number = Date.now(),
  algorithmV4Enabled: boolean = false,
): ScheduleResult => {
  if (algorithmV4Enabled) {
    return scheduleSpellingV4(current, outcome, now);
  }

  const prevReviewCount = current?.reviewCount ?? 0;
  const prevIntervalDays = current?.intervalDays ?? 0;
  const prevRepetitions = current?.repetitions ?? 0;
  const learningPhase = isLearningPhase(current);
  let ease = updateEase(current?.easeFactor ?? DEFAULT_EASE, outcome, MIN_EASE_SPELLING);

  if (outcome === SPELLING_OUTCOME_FAILED) {
    ease = Math.max(MIN_EASE_SPELLING, ease - 0.2);
    return {
      repetitions: 0,
      intervalDays: V3_MIN_INTERVAL_DAYS,
      easeFactor: ease,
      nextReviewTime: nextReviewTimeByDays(V3_MIN_INTERVAL_DAYS, now),
      status: PROGRESS_STATUS_LEARNING,
    };
  }

  let intervalDays: number;
  if (learningPhase) {
    intervalDays =
      outcome === SPELLING_OUTCOME_RETRY_SUCCESS ? V3_MIN_INTERVAL_DAYS : V3_GOOD_INTERVAL_DAYS;
  } else if (outcome === SPELLING_OUTCOME_RETRY_SUCCESS) {
    intervalDays = conservativeInterval(Math.max(1, prevIntervalDays), ease);
  } else if (outcome === SPELLING_OUTCOME_HINTED) {
    intervalDays = standardGoodInterval(Math.max(1, prevIntervalDays), ease);
  } else {
    const base = standardGoodInterval(Math.max(1, prevIntervalDays), ease);
    intervalDays = Math.max(1, Math.round(base * 1.1));
  }

  const effectiveReviewCount = prevReviewCount + 1;
  return {
    repetitions: prevRepetitions + 1,
    intervalDays,
    easeFactor: ease,
    nextReviewTime: nextReviewTimeByInterval(now, intervalDays),
    status: resolveStatus(intervalDays, effectiveReviewCount, ease, false),
  };
};

const scheduleSpellingV4 = (
  current: Progress | null,
  outcome: SpellingOutcome,
  now: number,
): ScheduleResult => {
  const prevReviewCount = current?.reviewCount ?? 0;
  const prevIntervalDays = current?.intervalDays ?? 0;
  const prevRepetitions = current?.repetitions ?? 0;
  const learningPhase = isLearningPhase(current);
  let ease = updateEase(current?.easeFactor ?? DEFAULT_EASE, outcome, MIN_EASE_SPELLING);

  if (outcome === SPELLING_OUTCOME_FAILED) {
    ease = Math.max(MIN_EASE_SPELLING, ease - 0.2);
    return {
      repetitions: 0,
      intervalDays: 0,
      easeFactor: ease,
      nextReviewTime: now + ONE_MINUTE,
      status: PROGRESS_STATUS_LEARNING,
    };
  }

  let intervalDays: number;
  if (learningPhase) {
    intervalDays =
      outcome === SPELLING_OUTCOME_RETRY_SUCCESS ? 0 : nextEbbinghausStep(prevIntervalDays);
  } else {
    const oldInterval = Math.max(1, prevIntervalDays);
    let rawInterval: number;
    if (outcome === SPELLING_OUTCOME_RETRY_SUCCESS) {
      rawInterval = conservativeInterval(oldInterval, ease);
    } else if (outcome === SPELLING_OUTCOME_HINTED) {
      rawInterval = standardGoodInterval(oldInterval, ease);
    } else if (outcome === SPELLING_OUTCOME_PERFECT) {
      rawInterval = Math.max(1, Math.round(standardGoodInterval(oldInterval, ease) * 1.1));
    } else {
      rawInterval = 1;
    }
    intervalDays = closestEbbinghausStep(rawInterval);
  }

  if (intervalDays > 1 && reviewedWithinHours(current, now, SHORT_TERM_REVIEW_DECAY_HOURS)) {
    intervalDays = 1;
  }

  const effectiveReviewCount = prevReviewCount + 1;
  return {
    repetitions: prevRepetitions + 1,
    intervalDays,
    easeFactor: ease,
    nextReviewTime: nextReviewTimeByInterval(now, intervalDays),
    status: resolveStatus(intervalDays, effectiveReviewCount, ease, true),
  };
};

export const evaluateRecognitionByResponseTime = (
  rating: StudyRating,
  responseTimeMs: number,
  algorithmV4Enabled: boolean,
  thresholdMs: number,
): StudyRating => {
  if (!algorithmV4Enabled) return rating;
  if (responseTimeMs <= thresholdMs) return rating;
  if (rating === STUDY_RATING_GOOD) return STUDY_RATING_HARD;
  if (rating === STUDY_RATING_HARD) return STUDY_RATING_AGAIN;
  return STUDY_RATING_AGAIN;
};
